activerecord/lib/active_record/query_cache.rb

```diff
@@ -44,8 +44,9 @@ def self.install_executor_hooks(executor = ActiveSupport::Executor)
       executor.register_hook(self)
 
       executor.to_complete do
-        # FIXME: This should be skipped when env['rack.test']
-        ActiveRecord::Base.clear_active_connections!
+        unless ActiveRecord::Base.connection.transaction_open?
+          ActiveRecord::Base.clear_active_connections!
+        end
```

Comment:
Given that the discussion here is about the ownership of the connection for the current thread, and whether we should try to release it, the pool shouldn't be involved at all here. Clearing unused connections from other threads is a side effect of checking the current connection in, not the main goal.